src/common/comp/brute_force.cc

```diff
@@ -352,7 +385,14 @@ BruteForce::RangeSearch(const DataSetPtr base_dataset, const DataSetPtr query_da
                     if (!bitset.empty() && bitset.test(j)) {
                         continue;
                     }
-                    auto dist = cur_query->dot(xb_sparse[j]);
+                    float row_sum = 0;
+                    if (is_bm25) {
+                        for (size_t k = 0; k < xb_sparse[j].size(); ++k) {
+                            auto [d, v] = xb_sparse[j][k];
+                            row_sum += v;
+                        }
+                    }
+                    auto dist = cur_query->dot(xb_sparse[j], sparse_computer, row_sum);
```

Comment:
Please move this code inside `SparseRow` to have the caller to use it as follows:
```
float dist = 0;
if (is_bm25) {
    dist = cur_query->dot_sparse(xb_sparse[j], sparse_computer);
} else {
    dist = cur_query->dot(xb_sparse[j]);
}
```
The code from the change belongs to `SparseRow` class.

Reply:
the reason why I made the caller to compute and pass in the row sum: 

* we don't need to compute the row sum for IP metric, thus I don't want to make `row_sum` an always computed member of `SparseRow`. 
* if we want to do some "lazy compute row sum and store in SparseRow" thing, we need to protect from concurrent row sum update which might be an overkill? also `dot` is a const method which makes it harder
* with bm25 on index we can precompute and cache the row sums, so that in `InvertedIndex::refine_and_collect` we pass the cached row sum to `dot`
* in the case of brute force, the incoming Dataset is a const one and we cannot modify it(create/update cached row sum)

if we still want to have the `SparseRow` to own the row sum computation logic:

```cpp
// dot_sparse to accept an optional row sum:
float dot_sparse(other, computer, std::optional<T> other_row_sum = std::nullopt) const {
    T row_sum = other_sum.has_value() ? other_sum.value() : row_sum();
    ...
}

// computes row sum on the fly, does not cache
float row_sum() const {}
```

WDYT?

Reply:
your arguments totally make sense.